ports/atmel-samd/common-hal/watchdog/WatchDogTimer.c

```diff
@@ -0,0 +1,312 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+
+#include "common-hal/watchdog/WatchDogTimer.h"
+
+#include "shared-bindings/watchdog/__init__.h"
+#include "shared-bindings/microcontroller/__init__.h"
+
+#include <sam.h>
+
+STATIC bool _watchdog_samd_initialized;
+
+STATIC void _watchdog_samd_reset(void) {
+  // Write the watchdog clear key value (0xA5) to the watchdog
+  // clear register to clear the watchdog timer and reset it.
+#ifdef SAMD51
+  while (WDT->SYNCBUSY.reg)
+    ;
+#else
+  while (WDT->STATUS.bit.SYNCBUSY)
+    ;
+#endif
+  WDT->CLEAR.reg = WDT_CLEAR_CLEAR_KEY;
+}
+
+STATIC void _watchdog_samd_disable(void) {
+#ifdef SAMD51
+  WDT->CTRLA.bit.ENABLE = 0;
+  while (WDT->SYNCBUSY.reg)
+    ;
+#else
+  WDT->CTRL.bit.ENABLE = 0;
+  while (WDT->STATUS.bit.SYNCBUSY)
+    ;
+#endif
+}
+
+STATIC void _watchdog_samd_initialize_wdt(void) {
+  // One-time initialization of watchdog timer.
+  // Insights from rickrlh and rbrucemtl in Arduino forum!
+
+#ifdef SAMD51
+  // SAMD51 WDT uses OSCULP32k as input clock now
+  // section: 20.5.3
+  OSC32KCTRL->OSCULP32K.bit.EN1K = 1;  // Enable out 1K (for WDT)
+  OSC32KCTRL->OSCULP32K.bit.EN32K = 0; // Disable out 32K
+
+  // Enable WDT early-warning interrupt
+  NVIC_DisableIRQ(WDT_IRQn);
+  NVIC_ClearPendingIRQ(WDT_IRQn);
+  NVIC_SetPriority(WDT_IRQn, 0); // Top priority
+  NVIC_EnableIRQ(WDT_IRQn);
+
+  while (WDT->SYNCBUSY.reg)
+    ;
+
+  USB->DEVICE.CTRLA.bit.ENABLE = 0; // Disable the USB peripheral
+  while (USB->DEVICE.SYNCBUSY.bit.ENABLE)
+    ;                                 // Wait for synchronization
+  USB->DEVICE.CTRLA.bit.RUNSTDBY = 0; // Deactivate run on standby
+  USB->DEVICE.CTRLA.bit.ENABLE = 1;   // Enable the USB peripheral
+  while (USB->DEVICE.SYNCBUSY.bit.ENABLE)
+    ; // Wait for synchronization
+#else
+  // Generic clock generator 2, divisor = 32 (2^(DIV+1))
+  GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(4);
+  // Enable clock generator 2 using low-power 32KHz oscillator.
+  // With /32 divisor above, this yields 1024Hz(ish) clock.
+  GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2) | GCLK_GENCTRL_GENEN |
+                      GCLK_GENCTRL_SRC_OSCULP32K | GCLK_GENCTRL_DIVSEL;
+  while (GCLK->STATUS.bit.SYNCBUSY)
+    ;
+  // WDT clock = clock gen 2
+  GCLK->CLKCTRL.reg =
+      GCLK_CLKCTRL_ID_WDT | GCLK_CLKCTRL_CLKEN | GCLK_CLKCTRL_GEN_GCLK2;
```

Comment:
We manage the clocks differently. `GCLK2` could well be used by something else. See `ports/atmel-samd/peripherals`, which is a submodule connected to https://github.com/adafruit/samd-peripherals. You may need to add functionality to that library, or maybe not.

It probably makes sense to move some or all of the samd-specifc code here into that library. You could then provide a uniform API and have the library take care of the SAMD21/51 differences.

Reply:
This is more or less greek to me sadly.  I've more or less copy and pasted the code from the sleepydog repo without much thought.  If this is a blocker, I'll take the time to grok it.